### PR TITLE
Update Resource Forms to Enhance Validation

### LINKED ISSUE #9053: Resource Update Allows Negative Quantity and Empty Description Fields
**Bug Description**
The quantity field allows negative values during resource updates, and the mandatory description field can be left empty, which should not be allowed.

---

**Steps to Reproduce**
1. Navigate to **resource/list**.
2. Select **All Details** of any resource.
3. Click on **Update Status/Details**.
4. Enter a negative value in the **Required Quantity** field and clear the **Description** field.
5. Attempt to update the resource and observe the behavior.

---

**Expected Behavior**
- The **Required Quantity** field should not accept negative values.
- The **Description** field should be mandatory, preventing updates if left empty.

---

**Screenshots**
Screenshot showing error

---

**Environment**
- **OS:** Windows
- **Browser:** Chrome
- **Version:** 130.0.6723.117

### PR DESCRIPTION
**Pull request by JavidSumra**
## Proposed Changes

- Fixes #9053
- Prevents negative values in the **Required Quantity** field during resource updates.
- Enforces a non-empty **Description** field as mandatory for updates.

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network) (if applicable).
- [x] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [ ] Completion of QA

## Summary by CodeRabbit

- **New Features**
	- Enhanced resource creation component with improved form handling and validation.
	- New localization strings added for better user guidance and error handling.

- **Bug Fixes**
	- Updated input field selectors to align with the current HTML structure, ensuring accurate data capture during resource requests.

- **Tests**
	- Modified test suite to include new functionality for updating resource status.

- **Chores**
	- Refactored routing logic to streamline resource update process.

### CODE REVIEW

**Comment by nihal467** [comment]
image

- [ ] the PR is not solving the issue @JavidSumra

**Comment by JavidSumra** [comment]
Have you updated the value or is it already present there?
cc @nihal467

**Comment by nihal467** [comment]
@JavidSumra i updated the value

File: src/components/Form/FormFields/TextFormField.tsx
Comment (rithviknishad) [inline src/components/Form/FormFields/TextFormField.tsx]: this is given by `DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>` right? Do we need to redefine it here?

File: src/components/Form/FormFields/TextFormField.tsx
Comment (rithviknishad) [inline src/components/Form/FormFields/TextFormField.tsx]: is this also needed? `{...props}` in line 91 already handles it right.

```suggestion
```

File: src/Utils/utils.ts
Comment (rithviknishad) [inline src/Utils/utils.ts]: `handleNegativeValue` does not convey what the function does. Like what does it handle exactly? Fn. name can be improved.

```suggestion
export const sanitizePositiveNumberValue = (
```

File: src/Utils/utils.ts
Comment (rithviknishad) [inline src/Utils/utils.ts]: this just clears the input everytime a special character is pressed. Instead of clearing, I'd recommend cleaning the non numeric characters so that user doesn't loose what they've typed.

Eg: User types `12)`, instead of clearing when i type `)`, it should reset to just `12`

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/Utils/utils.ts
Comment (JavidSumra) [reply src/Utils/utils.ts]: Hey @rithviknishad, I updated the function logic, but it's clearing the value when entering '+/-' after any digit. Do you have any suggestions to improve the logic?
```js
export const sanitizeNumberInput = (
  event: React.FormEvent<HTMLInputElement>,
): void => {
  const input = event.currentTarget;
  input.value = input.value.replace(/[^0-9]/g, "");
};
```

**Comment by nihal467** [comment]
image

- [ ] We can create a new resource request with a required quantity of 0. However, after creation, when we open the update resource page, a validation error is thrown. If the value is restricted, the error should be thrown during creation itself.

**Comment by JavidSumra** [comment]
Hey @nihal467 can you please add `needs-testing` label on this PR?

**Comment by nihal467** [comment]
image

- [ ] when we create a new resource request, the approved quantity is having the same error mentioned before. 

@JavidSumra before requesting a review, make sure the PR is working well

**Comment by nihal467** [comment]
image

- [ ] When you are creating a new resource request, without even approving a quantity its showing 1

**Comment by JavidSumra** [comment]
> image
> 
> - [ ] When you are creating a new resource request, without even approving a quantity its showing 1

I added 1 as a default value for approving quantity (my bad) shall I set it to 0?
cc @nihal467

**Comment by nihal467** [comment]
> > image
> > 
> > * [ ]  When you are creating a new resource request, without even approving a quantity its showing 1
> 
> I added 1 as a default value for approving quantity (my bad) shall I set it to 0? cc @nihal467

Let’s keep it as null if the value is not submitted by the users, instead of defaulting it to 1 or 0.

**Comment by JavidSumra** [comment]
Hey @nihal467  can you please update your review over here done with requested changes

File: src/components/Resource/ResourceDetailsUpdate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceDetailsUpdate.tsx]: change the return in validateForm instead 👍

File: src/components/Resource/ResourceCreate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceCreate.tsx]: Let's not change what the user enters, instead trigger the error message in validateForm.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

File: src/components/Resource/ResourceCreate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceCreate.tsx]: Remove the console log

File: src/components/Resource/ResourceCreate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceCreate.tsx]: Use the `errorText` added above.

Since you are at it, add translations for all the error texts.

File: src/components/Resource/ResourceDetailsUpdate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceDetailsUpdate.tsx]: While this works, it isn't the best option readability wise (esp if we add any new fields). Let's switch to case statements as we have in other components.

That also allows to add any extra checks in the future, if we need different error messages.

File: src/components/Resource/ResourceDetailsUpdate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceDetailsUpdate.tsx]: Translations 👍

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

**Review by Jacobjeevan — COMMENTED** [review]
LGTM so far.

Hmm, go ahead and convert to using Form component (src/components/Form) here, should make state management easier as well.

See if you can create a ResourceForm component that could be used in both pages (AssetCreate does this). You can use #9192 as an example.

**Comment by nihal467** [comment]
@Jacobjeevan can you review the PR

File: src/components/Resource/ResourceCreate.tsx
Comment (Jacobjeevan) [inline src/components/Resource/ResourceCreate.tsx]: See if you can use ResourceModel instead of defining another data structure here.

Probably will have to use Partial<ResourceModel>

File: src/Routers/routes/ResourceRoutes.tsx
Comment (Jacobjeevan) [reply src/Routers/routes/ResourceRoutes.tsx]: Make facilityId optional in ResourceCreate instead.

**Review by Jacobjeevan — CHANGES_REQUESTED** [review]

**Comment by JavidSumra** [comment]
@nihal467 need your review

**Comment by Jacobjeevan** [comment]
Resources was cleaned up in #9439. Hold this for now (you are free to refer to it and make your changes based on those before encountering any merge conflicts).

**Comment by rithviknishad** [comment]
no longer relevant in new architecture

**Pull request closed without merging**
